src/call_context.rs

```diff
@@ -382,25 +392,31 @@ impl<'a> CallContext<'a> {
         self.block_height
     }
 
-    pub fn gas_meter(&self) -> &GasMeter {
-        &self.top().gas_meter
+    pub fn gas_meter(&mut self) -> Result<&GasMeter, VMError> {
+        let stack = &mut self.top_mut()?;
+        let instance = &stack.instance;
+        let gas_meter = &mut stack.gas_meter;
+
+        gas_meter.update(instance, 0)?;
+
+        Ok(&self.top()?.gas_meter)
     }
 
-    pub fn gas_meter_mut(&mut self) -> &mut GasMeter {
-        &mut self.stack.last_mut().expect("Invalid Stack").gas_meter
+    pub fn top(&self) -> Result<&StackFrame, VMError> {
+        self.stack.last().ok_or(VMError::EmptyStack)
     }
 
-    pub fn top(&self) -> &StackFrame {
-        self.stack.last().expect("Invalid stack")
+    pub fn top_mut(&mut self) -> Result<&mut StackFrame, VMError> {
+        self.stack.last_mut().ok_or(VMError::EmptyStack)
     }
 
-    pub fn callee(&self) -> &ContractId {
-        &self.top().callee
+    pub fn callee(&self) -> Result<&ContractId, VMError> {
```

Comment:
I am not sure it is worth introducing explicit error handling here, as this function may fail only if there is no stack, which IMO is better handled by assertions or expectations (hence - as before). The same remark applies to the `caller`, `top`, and `top_mut` methods.

Reply:
I agree. It's clearly the type of programmer error that shouldn't be propagated to the user. Removed.